### ISSUE TITLE
Fix flakiness in CI by addressing root cause(s)

### ISSUE THREAD

**Issue by jdegoes** [zio/zio-http#2400]
Currently CI is extremely flaky with many hangs, especially around streaming tests. Some efforts were spent in diagnosing this problem and it seems part of the problem is hangup in layer finalization that occurs when servers (and clients) are shut down. @vigoo confirmed this was not due to the code that waits for all active connections to be gracefully shut down. It's known to be racy in that adding delays into finalization seems to reduce incidence of hanging.

This ticket will be fixed when layer disconnection is not required, when layers finalize properly and in a timely fashion, when all tests (including streaming) can be re-enabled on CI without causing hanging.

If you have worked on this issue, please post as much diagnostic information as you have so whoever takes this issue on can do so with our collective knowledge.

**Comment by adamgfraser** [comment]
Will take a look at this.

**Comment by ayusham001** [comment]
/attempt #2400



<details id="algora-options">
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/cllp895w90002l10fq8iei13z/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by adamgfraser** [comment]
I think this issue may be more actionable if we break it up into smaller pieces.

We had issues with tests hanging and layer finalization,. I believe those are resolved by #2407. While it is challenging to prove a negative, I have not observed any warnings regarding layer disconnection since fixing the test scopes in that PR over many CI runs.

With that resolved, we have the below remaining ignored tests.

**ClientHttpsSpec**

There is one ignored test here that is trying to assert that you get a bad request response back for a particular request. This just fails if enabled because the response is okay. So there is some specific functionality here that needs to be fixed.

**WebSocketSpec**

There is one ignored test here involving multiple web socket upgrades. If enabled, this test fails with the following message:

```
Exception in thread "zio-fiber-419589,408793" java.lang.IllegalArgumentException: Duplicate handler name: WEB_SOCKET_CLIENT_PROTOCOL_HANDLER
```

This indicates that we are trying to add the same handler twice to the same pipeline. We should be creating a new channel and pipeline for each request but apparently this is not happening so this needs to be investigated.

**ClientStreamingSpec**

There are three tests that time out in CI here. All of them involve decoding of form data with property based tests, so if is possible that these tests are just taking too long to run but needs to be investigated further.

**EndpointRoundTripSpec**

There are two tests here that are failing in CI involving multi part input with the endpoint DSL. They time out and produce an error message that looks like this:

```
              Console IO Produced by Test         
    | Failed POST http://localhost:42563/test 50036ms: Exception in thread "zio-fiber-18467" zio.Cause$Fail$$anon$17: Fail(io.netty.handler.timeout.ReadTimeoutException,Stack trace for thread "zio-fiber-18467":
	at zio.http.ZClient.request(ZClient.scala:165)
	at zio.http.endpoint.EndpointRoundtripSpec.testEndpointZIO(EndpointRoundtripSpec.scala:94)
	at zio.http.endpoint.EndpointRoundtripSpec.testEndpointZIO(EndpointRoundtripSpec.scala:96)
	at zio.http.endpoint.EndpointRoundtripSpec.spec(EndpointRoundtripSpec.scala:294))
	at zio.http.ZClient.request(ZClient.scala:165)
	at zio.http.endpoint.EndpointRoundtripSpec.testEndpointZIO(EndpointRoundtripSpec.scala:94)
	at zio.http.endpoint.EndpointRoundtripSpec.testEndpointZIO(EndpointRoundtripSpec.scala:96)
	at zio.http.endpoint.EndpointRoundtripSpec.spec(EndpointRoundtripSpec.scala:294)
```

**SSLSpec**

These tests appear to just be misspecified and attempt to construct a server before running tests but do `Server.serve` which never returns so the tests never get a chance to execute.
